src/sources/kafka.rs

```diff
@@ -236,25 +238,43 @@ async fn kafka_source(
     acknowledgements: bool,
 ) -> Result<(), ()> {
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
+    let (mut finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new_without_shutdown(acknowledgements);
     let mut stream = consumer.stream();
     let keys = Keys::from(log_schema(), &config);
 
+    let mut shutting_down = false;
     loop {
         tokio::select! {
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+            _ = &mut shutdown => {
+                if let Ok(topics) = consumer.subscription() {
+                    if let Err(error) = consumer.pause(&topics) {
+                        // Might not be the best error type for this :/
+                        emit!(KafkaReadError { error });
                     }
                 }
+                shutting_down = true;
+
+                // If we have an ack stream, drop the sender end and allow acks to drain
+                match finalizer {
+                    Some(_) => drop(finalizer.take()),
+                    _ => break
+                }
             },
-            message = stream.next() => match message {
-                None => break,  // WHY?
+            entry = ack_stream.next() => match entry {
+                Some((status, entry)) => {
+                    if status == BatchStatus::Delivered {
+                        if let Err(error) =
+                            consumer.store_offset(&entry.topic, entry.partition, entry.offset)
+                            {
+                                emit!(KafkaOffsetUpdateError { error });
+                            }
+                    }
+                },
+                None => break,
+            },
+            message = stream.next(), if !shutting_down => match message {
+                None => break,  // No more messages...but is this the correct way to handle it?
```

Comment:
Can we treat this the same way as a shutdown, by setting `shutting_down = true`?

Reply:
I think we might want to run the shutdown block here actually. Although, I'm not sure what will cause this branch to match-- there doesn't seem to be a way to close or otherwise end a `MessageStream` from the client side so this might only end when the consumer goes out of scope?

Reply:
Is it correct that since rdkafka's `MessageStream.poll_next` never returns `Poll::Ready(None)` ([source](https://docs.rs/rdkafka/0.28.0/src/rdkafka/consumer/stream_consumer.rs.html#119-140)), this branch is unreachable?